osf/external/spam/tasks.py

```diff
@@ -0,0 +1,20 @@
+from framework.celery_tasks import app as celery_app
+from django.db import transaction
+
+@celery_app.task()
+def reclassify_domain_references(notable_domain_id):
+    from osf.models.notable_domain import DomainReference, NotableDomain
+    domain = NotableDomain.load(notable_domain_id)
+    references = DomainReference.objects.filter(domain=domain)
+    with transaction.atomic():
+        for item in references:
+            if domain.note == NotableDomain.Note.EXCLUDE_FROM_ACCOUNT_CREATION_AND_CONTENT:
+                item.referrer.confirm_spam(save=True)
+                item.is_triaged = True
+            if domain.note == NotableDomain.Note.UNKNOWN:
+                item.referrer.remove_flag(save=True)
```

Comment:
Un-SPAMing is...tricky. This is assuming that this particular domain was the reason something was marked as spam. But it could have had multiple SPAM domains or been flagged by Oopspam/akismet for some other reason.

We need some sort of `should_unspam` function that checks to make sure that there aren't any other references to `SPAM`y domains, at the very least. We probably also need to do something like add a `spam_reason` param to the node log that we pass when spamming because of a spamm-y domain. We may *also* need to short-circuit `confirm_spam` to exit early if the resource is already *SPAM*ed.

Finally, `remove_flag` only works in the case where spam status if `FLAGGED` (and it doesn't undo all of the deletion/privacy-setting bits). So we probably need to extract some `confirm_ham` logic into a more generic `unspam` function that doesn't then give the resource a free pass. Blech.

There's probably a whole extra ticket (or two) worth of work here that we didn't quite think through.

Reply:
I misunderstood what `remove_flag` means. It seems that the spam workflows was initially designed so that nothing is moved from `SpamStatus.UNKNOWN` to `SpamStatus.SPAM` without being first flagged as `SpamStatus.FLAGGED`. It was the OSF Admin's responsibility to move something from `SpamStatus.FLAGGED` to either `SpamStatus.SPAM` or `SpamStatus.HAM`.

Therefore, it seems we are foregoing that workflow and skip the `FLAGGED` status if the resourced is tied to a `NotableDomain`. If so, then we do need to add some logic to tell whether something could be unspammed. Or instead of jumping from `UNKNOWN` to `SPAM`, we flag everything upon reclassification and let prod folks do the dirty work?

Reply:
Sorry, maybe tried to cram too much into one comment. To be clear, there are two issues:

One: we don't know if this domain is the reason this resource is spammed (or flagged if we go that route). There could be multiple spam flags and/or it could have spam domains and have been flagged by the external spam detector. Just because we triage one domain from SPAM to not doesn't mean we should automatically un-spam/flag *everything* that references the formerly-spammy domain.

Two: The logic being invoked only works on things that have been flagged, not actively SPAMed.

We can talk at our meeting today about FLAGing things instead of straight-up SPAMing them, but that doesn't resolve the first problem, which is the source of most of the required extra work.